rclrs/src/timer.rs

```diff
@@ -0,0 +1,416 @@
+                // SAFETY:
+                // * The rcl_timer is zero-initialized as mandated by this function.
+                // * The rcl_clock is kept alive by the Clock within TimerHandle because it is
+                //   a dependency of the timer.
+                // * The rcl_context is kept alive by the ContextHandle within TimerHandle because
+                //   it is a dependency of the timer.
+                // * The period is copied into this function so it can be dropped afterwards.
+                // * The callback is None / nullptr so doesn't need to be kept alive.
+                // * The entity lifecycle mutex is locked to protect against the risk of global
+                //   variables in the rmw implementation being unsafely modified during initialization.
+                rcl_timer_init(
+                    &mut rcl_timer,
+                    &mut *rcl_clock,
+                    &mut *rcl_context,
+                    period_nanos,
+                    None,
+                    allocator,
+                )
+                .ok()?;
+            }
+        }
+
+        Ok(Self {
+            callback,
+            handle: TimerHandle {
+                rcl_timer: Mutex::new(rcl_timer),
+                _clock: clock,
+                _context_handle: context_handle,
+                in_use_by_wait_set: Arc::new(AtomicBool::new(false)),
+            },
+        })
+    }
+
+    /// Calculates if the timer is ready to be called.
+    /// Returns true if the timer is due or past due to be called.
+    /// Returns false if the timer is not yet due or has been canceled.
+    pub fn is_ready(&self) -> bool {
+        let mut timer = self.handle.lock();
+        let mut is_ready = false;
+        // SAFETY:
+        // * The timer is initialized, which is guaranteed by the constructor.
+        // * The is_ready pointer is allocated on the stack and is valid for the duration of this function.
+        let ret = unsafe { rcl_timer_is_ready(&mut *timer, &mut is_ready) };
+
+        // rcl_timer_is_ready should only error if incorrect arguments are given or something isn't initialised,
+        // both of which we control in this function.
+        debug_assert_eq!(RclReturnCode::try_from(ret).unwrap(), RclReturnCode::Ok);
+
+        is_ready
+    }
+
+    /// Get the time until the next call of the timer is due. Saturates to 0 if the timer is ready.
+    /// Returns [`RclReturnCode::TimerCanceled`] as an error if the timer has already been canceled.
+    pub fn time_until_next_call(&self) -> Result<Duration, RclrsError> {
+        let mut timer = self.handle.lock();
+        let mut remaining_time = 0;
+        // SAFETY:
+        // * The timer is initialized, which is guaranteed by the constructor.
+        // * The remaining_time pointer is allocated on the stack and is valid for the duration of this function.
+        unsafe {
+            rcl_timer_get_time_until_next_call(&mut *timer, &mut remaining_time).ok()?;
```

Comment:
As per: https://github.com/geoff-imdex/ros2_rust/actions/runs/11564633206/job/32190192780#step:11:395
```suggestion
            rcl_timer_get_time_until_next_call(& *timer, &mut remaining_time).ok()?;
```

Reply:
Good catch, fixed.